test/uk/gov/hmrc/transitmovementsrouter/services/XmlParserSpec.scala

```diff
@@ -62,62 +61,87 @@ class XmlParserSpec extends AnyFreeSpec with TestActorSystem with Matchers {
+            _.right.get mustBe OfficeOfDeparture("GB6789")
+          }
+        }
+
+        s"when provided with a missing OfficeOfDeparture node in ${messageType.code} message it returns NoElementFound" in {
+          val stream       = createParsingEventStream(messageWithoutOfficeOfDeparture(messageType.rootNode))
+          val parsedResult = stream.via(XmlParser.officeOfDepartureExtractor(MessageType.DeclarationData)).runWith(Sink.head)
+
+          whenReady(parsedResult) {
+            _.mustBe(Left(NoElementFound("referenceNumber")))
+          }
+        }
+
+        s"when provided with a missing ReferenceNumber node in ${messageType.code} message it returns NoElementFound" in {
+          val stream       = createParsingEventStream(messageWithoutRefNumber(messageType.rootNode))
+          val parsedResult = stream.via(XmlParser.officeOfDepartureExtractor(MessageType.DeclarationData)).runWith(Sink.head)
+
+          whenReady(parsedResult) {
+            _.mustBe(Left(NoElementFound("referenceNumber")))
+          }
+        }
       }
-    }
 
-    "when provided with a missing OfficeOfDeparture node" in {
-      val stream       = createParsingEventStream(cc015cNoOfficeOfDeparture)
-      val parsedResult = stream.via(XmlParser.officeOfDepartureExtractor).runWith(Sink.head)
+      s"MessageSenderElement parser should add the messageSender node with the movement Id in ${messageType.code} message" in new Setup {
 
-      whenReady(parsedResult) {
-        _.mustBe(Left(NoElementFound("referenceNumber")))
-      }
-    }
+        val stream = createStream(messageOfficeOfDepartureGb(messageType.rootNode))
 
-    "when provided with a missing ReferenceNumber node" in {
-      val stream       = createParsingEventStream(cc015cNoRefNumber)
-      val parsedResult = stream.via(XmlParser.officeOfDepartureExtractor).runWith(Sink.head)
+        val parsedResult = stream
+          .via(XmlParsing.parser)
+          .via(XmlParser.messageSenderWriter(messageType, MessageSender("MVN123456789-MSG987654321"))) // testing this
+          .via(XmlWriting.writer)
+          .fold(ByteString())(_ ++ _)
+          .map(_.utf8String)
+          .runWith(Sink.head)
 
-      whenReady(parsedResult) {
-        _.mustBe(Left(NoElementFound("referenceNumber")))
+        whenReady(parsedResult) {
+          result =>
+            XML.loadString(result) shouldBe messageWithExpectedMessageSender(messageType.rootNode)
+        }
       }
-    }
   }
 
-  "MessageSenderElement parser should add the messageSender node with the movement Id" in new Setup {
+  trait Setup {
 
-    val stream = createStream(cc015cNoMessageSenderNode)
+    def messageOfficeOfDepartureGb(messageTypeNode: String): NodeSeq = {
+      val strMessage =
+        s"<$messageTypeNode><preparationDateAndTime>2022-03-28T09:15:12</preparationDateAndTime><CustomsOfficeOfDeparture><referenceNumber>GB6789</referenceNumber></CustomsOfficeOfDeparture></$messageTypeNode>"
```

Comment:
These nodes aren't technically correct with regards to what you'll receive. I think your actual XML parsing code is right though -- it's all to do with the namespaces.

The XML root node should look something like:

```xml
<ncts:CC014C PhaseID="NCTS5.0" xmlns:ncts="http://ncts.dgtaxud.ec"></ncts:CC014C>
```

We should update all tests to use this.